parla/task_runtime.py

```diff
@@ -349,118 +347,163 @@ def result(self):
+                        # notify successors and make them wait for that event,
+                        # not Parla task completion.
+                        self._notify_successors_mutex(events)
+                    env.sync_events()
+                task_state = task_state or TaskCompleted(None)
+            except Exception as e:
+                task_state = TaskException(e)
+                logger.exception("Exception in task")
+            finally:
+                logger.info("Finally for task %r", self)
+
+                ctx = get_scheduler_context()
+                self._post_run(ctx)
+
+                # Protect the case that it notifies successors and
+                # any successor task is spawned before setting state.
+                with self._mutex:
+                    # Regardless of the previous notification,
+                    # (So, before leaving the current run(), the above)
+                    # it should notify successors since
+                    # new successors could be added after the above
+                    # notifications, while other devices are running
+                    # their kernels asynchronously.
+                    self._notify_successors()
+                    self._set_state(task_state)
+
+        except Exception as e:
+            logger.exception("Task %r: Exception in task handling", self)
+            raise e
+
+
     def set_assigned(self):
         with self._mutex:
-            self.assigned = True
+            self._assigned = True
 
     def is_assigned(self):
         with self._mutex:
-            if self.assigned:
-                return True
-            else:
-                return False
+            return self._assigned
 
-    def _set_dependencies(self, dependencies):
-        self._dependencies = dependencies
-        self._remaining_dependencies = len(dependencies)
-        for dep in dependencies:
+    @property
+    def predecessors(self) -> Tuple["Task"]:
+        return tuple(self._predecessors)
+
+    @predecessors.setter
+    def predecessors(self, predecessors: Collection['Task']):
+        self._predecessors: List[Task] = list(predecessors)
+        self._num_blocking_predecessors: int = len(predecessors)
+        for dep in predecessors:
             # If a dependency is TaskID, not Task object,
             # it implies that it is not yet spawned.
             # Ignore it.
-            if isinstance(dep, TaskID):
-                continue
-            if not dep._add_dependee_mutex(self):
-                self._remaining_dependencies -= 1
-
-    def _set_dependencies_mutex(self, dependencies):
-        with self._mutex:
-            return self._set_dependencies(dependencies)
+            if not isinstance(dep, TaskID) and not dep._add_successor_mutex(self):
+                self._num_blocking_predecessors -= 1
 
-    def _check_remaining_dependencies(self):
-        if not self._remaining_dependencies and self.assigned:
+    def _maybe_ready_to_schedule(self):
```

Comment:
Can we make this `_check_remaining_predecessors` or something similar, and move the scheduling portion of it outside of this function or to a new function? `_maybe_ready_to_schedule` is unclear.

Reply:
agree

Reply:
It does more than checking; enqueuing could be a side effect. Agree to separate enqueuing.